### PR TITLE
Backend Auth (register and login)

### PR DESCRIPTION
**Pull request by Hakeemmidan**
In this pull request, I:
- Created `/auth/register` and `/auth/login` API routes (files listed under `routes/auth`)
  - Created validations for said routes (files listed under `validation/auth`)
- Created tests for said routes
  - [You don't need to run these as they're still a work in progress. I'm still trying to figure out how to use mocks.]

#### How to Test (on Postman)
- Run server first
- To test `/auth/register`:
  1. Pick `POST` as request type
  2. Place `localhost:3001/auth/register` in URL
  3. Go to `Body` and select `x-www-form-urlencoded`
  4. Place `name`, `email`, and `password` as keys and fill in values as needed

Successful registration should return to you an instance of the user.
Unsuccessful registration should return a customized message depending on what you're missing. Try having an empty password. It should return this to you:
```JSON
{
    "password": "Password must be at least 6 characters"
}
```

- To test `/auth/login`:
  1. Pick `POST` as request type
  2. Place `localhost:3001/auth/login` in URL
  3. Go to `Body` and select `x-www-form-urlencoded`
  4. Place `email` and `password` as keys and fill in values as needed

Successful login should return to you something like this:
```JSON
{
    "success": true,
    "token": "Bearer eyJhbGciOiJIUzI1NiI ... LfHtNFaa0GgQGqERh5eZ8dO0"
}
```
An unsuccessful registration should return a customized message depending on what you're missing.

---------

I wasn't sure how to format the JSON response. Feel free to open an issue about changing it if you think we should.

Please let me know if you have any questions
Thanks,
Hakeem = )

### CODE REVIEW

**Comment by Hakeemmidan** [comment]
Side note: I followed [this tutorial](https://blog.bitsrc.io/build-a-login-auth-app-with-mern-stack-part-1-c405048e3669)

**Comment by dotslash227** [comment]
I think it's great work. 

We should keep updating the readme with new API end-points, so that we keep on documenting the API end points. @skassam21 Should we document it as part of readme or use a different file or system?

**Review by dotslash227 — APPROVED** [review]

**Comment by Hakeemmidan** [comment]
I agree. I think that's a good idea. I'll wait for Shums to respond (or we can talk about it in today's stand-up) and I'll update it accordingly

File: server/routes/auth/login.route.js [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+
+  // Find user by email
+  User.findOne({ email }).then(user => {
+    // Check if user exists
+    if (!user) {
+      return res.status(400).json({ message: "Incorrect credentials" });
+    }
+
+    // Check password
+    bcrypt.compare(password, user.password).then(isMatch => {
```
Comment (skassam21) [inline server/routes/auth/login.route.js]: Maybe we want to move this into the model file - you can use parts of this tutorial - https://www.thepolyglotdeveloper.com/2019/02/hash-password-data-mongodb-mongoose-bcrypt/

File: server/routes/auth/register.route.js [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+    if (user) {
+      return res.status(400).json({ email: "Email already exists" });
+    } else {
+      const newUser = new User({
+        name: req.body.name,
+        email: req.body.email,
+        password: req.body.password
+      });
+      // Hash password before saving in database
+      bcrypt.genSalt(10, (err, salt) => {
```
Comment (skassam21) [inline server/routes/auth/register.route.js]: Same comment as above

File: server/routes/auth/register.route.js [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+        bcrypt.hash(newUser.password, salt, (err, hash) => {
+          if (err) throw err;
+          newUser.password = hash;
+          newUser
+            .save()
+            .then(user => {
+              delete user.password; // remove hashed password from response
+              return res.json(user);
```
Comment (skassam21) [inline server/routes/auth/register.route.js]: For signup, we will also want to generate a JWT, so we can login the user right away.

File: server/routes/auth/login.route.js [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+        jwt.sign(
+          payload,
+          keys.secretOrKey,
+          {
+            expiresIn: 31556926 // 1 year in seconds
+          },
+          (err, token) => {
+            res.json({
+              success: true,
+              token: "Bearer " + token
```
Comment (skassam21) [inline server/routes/auth/login.route.js]: We don't need to send this token anymore, as we will be using cookies (you can add that to the other PR)

File: server/routes/auth/login.route.js [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+    if (!user) {
+      return res.status(400).json({ message: "Incorrect credentials" });
+    }
+
+    // Check password
+    bcrypt.compare(password, user.password).then(isMatch => {
+      if (isMatch) {
+        const payload = {
+          id: user.id,
+          name: user.name
```
Comment (skassam21) [inline server/routes/auth/login.route.js]: We don't technically need the name in the payload

File: server/test/auth/register.spec.js [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+      password: "123",
+    };
+
+    it("it returns 200 status code", (done) => {
+      chai
+        .request(app)
+        .post("/register")
+        .send(user)
+        .end((err, res) => {
+          res.should.have.status(200);
```
Comment (skassam21) [inline server/test/auth/register.spec.js:43]: We should actually be checking if the information was added to the database. 

We also shouldn't be adding directly to our database in tests - take a look at:
- https://github.com/nodkz/mongodb-memory-server
- https://medium.com/javascript-in-plain-english/how-i-setup-unit-test-for-mongodb-using-jest-mongoose-103b772ee164 (Don't need to use jest, but use it to get the idea)

**Review by skassam21 — COMMENTED** [review]
Looks great! I would also take a look at this for how to structure an express back-end:
https://www.freecodecamp.org/news/how-to-write-a-production-ready-node-and-express-app-f214f0b17d8c/
@dotslash227 @tgnemecek @Hakeemmidan

**Review by skassam21 — APPROVED** [review]
Looks great!

**Pull request merged**